invokeai/app/services/shared/invocation_context.py

```diff
@@ -426,10 +430,161 @@ def search_by_attrs(
+          access_token: Optional access token for restricted resources.
+          timeout: Wait up to the indicated number of seconds before timing
+                   out long downloads.
+
+        Result:
+          Path of the downloaded model
+
+        May Raise:
+          HTTPError
+          TimeoutError
+        """
+        installer = self._services.model_manager.install
+        path: Path = installer.download_and_cache(
+            source=source,
+            access_token=access_token,
+            timeout=timeout,
+        )
+        return path
+
+    def load_ckpt_from_url(
+        self,
+        source: Union[str, AnyHttpUrl],
+        access_token: Optional[str] = None,
+        timeout: Optional[int] = 0,
+        loader: Optional[Callable[[Path], Dict[str | int, Any]]] = None,
+    ) -> LoadedModel:
+        """
+        Download, cache, and Load the model file located at the indicated URL.
+
+        This will check the model download cache for the model designated
+        by the provided URL and download it if needed using download_and_cache_ckpt().
+        It will then load the model into the RAM cache. If the optional loader
+        argument is provided, the loader will be invoked to load the model into
+        memory. Otherwise the method will call safetensors.torch.load_file() or
+        torch.load() as appropriate to the file suffix.
+
+        Be aware that the LoadedModel object will have a `config` attribute of None.
+
+        Args:
+          source: A URL or a string that can be converted in one. Repo_ids
+                  do not work here.
+          access_token: Optional access token for restricted resources.
+          timeout: Wait up to the indicated number of seconds before timing
+                   out long downloads.
+          loader: A Callable that expects a Path and returns a Dict[str|int, Any]
+
+        Returns:
+          A LoadedModel object.
+        """
+        ram_cache = self._services.model_manager.load.ram_cache
+        try:
+            return LoadedModel(_locker=ram_cache.get(key=str(source)))
+        except IndexError:
+            pass
+
+        def torch_load_file(checkpoint: Path) -> Dict[str | int, Any]:
+            scan_result = scan_file_path(checkpoint)
+            if scan_result.infected_files != 0:
+                raise Exception("The model at {checkpoint} is potentially infected by malware. Aborting load.")
+            return torch_load(path, map_location="cpu")
+
+        path = self.download_and_cache_ckpt(source, access_token, timeout)
+        if loader is None:
+            loader = (
+                torch_load_file
+                if path.suffix.endswith((".ckpt", ".pt", ".pth", ".bin"))
+                else lambda path: safetensors_load_file(path, device="cpu")
+            )
+
+        raw_model = loader(path)
+        ram_cache.put(key=str(source), model=raw_model)
+        return LoadedModel(_locker=ram_cache.get(key=str(source)))
```

Comment:
This loader pattern is clever. I think this is the first time we've enabled community nodes to register custom models with the model cache, right? That's pretty cool!

At some point we may want to formalize some sort of `CacheableModel` ABC interface that all models registered with the model cache have to implement. But, I think that can be left for future work.

Reply:
Nice idea. I'll make a stab at a `CacheableModel` in some future PR (if nobody does it first).

Reply:
On my list to do.